client/js/widgets/widget.js

```diff
@@ -304,8 +308,41 @@ export class Widget extends StateManaged {
     return css;
   }
 
+  cssAsText(css) {
+    if(typeof css == 'object') {
+      let cssText = '';
+      for(const key in css) {
+        if(typeof css[key] == 'object')
+          return this.cssToStylesheet(css);
+        cssText += `; ${key}: ${css[key]}`;
+      }
+      return cssText;
+    } else {
+      return css;
+    }
+  }
+
   cssProperties() {
-    return [ 'css', 'height', 'inheritChildZ', 'layer', 'width' ];
+    return [ 'css', 'height', 'inheritChildZ', 'layer', 'width' ].concat(this.propertiesUsedInCSS);
+  }
+
+  cssReplaceProperties(css) {
+    for(const match of css.matchAll(/\$\{PROPERTY ([A-Za-z0-9_-]+)\}/g)) {
+      css = css.replace(match[0], this.get(match[1]));
+      this.propertiesUsedInCSS.push(match[1]);
+    }
+    return css;
+  }
+
+  cssToStylesheet(css) {
+    const style = document.createElement('style');
+    style.id = `${this.id}STYLESHEET`;
+    for(const key in css) {
+      style.appendChild(document.createTextNode(`#${escapeCSS(this.id)}${key == 'default' ? '' : key} { ${this.cssReplaceProperties(this.cssAsText(css[key]))} }`));
+      $('head').appendChild(style);
+    }
+
+    return this.cssAsText(css.inline || '');
```

Comment:
Should probably also skip inline key when building the stylesheet.

Reply:
True.